File: server/internal/deployment/strategies/simple.go
package strategies

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"strings"
	"time"

	"github.com/docker/docker/api/types/container"
	"github.com/docker/docker/api/types/network"
	"github.com/docker/go-connections/nat"
	"github.com/puravida-software/bondi/server/internal/deployment/models"
	"github.com/puravida-software/bondi/server/internal/docker"
	"github.com/puravida-software/bondi/server/internal/docker/traefik"
)

const (
	_defaultNetworkName = "bondi-network"
	_serviceName        = "bondi-service"
	_traefikName        = "bondi-traefik"
)

var defaultNetworkingConfig = &network.NetworkingConfig{
	EndpointsConfig: map[string]*network.EndpointSettings{
		_defaultNetworkName: {},
	},
}

type SimpleDeployment struct {
	dockerClient docker.Client
}

func NewSimpleDeployment(dockerClient docker.Client) *SimpleDeployment {
	return &SimpleDeployment{dockerClient: dockerClient}
}

func (s *SimpleDeployment) Deploy(ctx context.Context, input *models.DeployInput) error {
	shouldRunTraefik := input.TraefikDomainName != nil && input.TraefikImage != nil && input.TraefikACMEEmail != nil
	if shouldRunTraefik {
		slog.Info("Creating network", "network_name", _defaultNetworkName)
		// Create a new network if it doesn't exist
		err := s.dockerClient.CreateNetworkIfNotExists(ctx, _defaultNetworkName)
		if err != nil {
			return fmt.Errorf("error creating network: %w", err)
		}

		traefikContainerID, err := s.runTraefik(ctx, input)
		if err != nil {
			return fmt.Errorf("error running Traefik: %w", err)
		}

		err = waitForTraefik(ctx, s.dockerClient, traefikContainerID)
		if err != nil {
			return fmt.Errorf("error waiting for Traefik to start: %w", err)
		}
	} else {
		slog.Info("Skipping network creation, Traefik is not enabled...")
	}

	// Get current container
	currentContainer, err := s.dockerClient.GetContainerByImageName(ctx, input.ImageName)
	if err != nil {
		return fmt.Errorf("error getting container: %w", err)
	}
	if currentContainer == nil {
		slog.Info("Container not found for image, assuming a fresh deployment...", "image_name", input.ImageName)
	} else {
		slog.Info("Container found for image", "container_id", currentContainer.ID)
	}

	// Stop current container
	if currentContainer != nil {
		slog.Info("Stopping current container", "container_id", currentContainer.ID)
		err = s.dockerClient.StopContainer(ctx, currentContainer.ID)
		if err != nil {
			return fmt.Errorf("error stopping container: %w", err)
		}

		slog.Info("Removing old image", "image_id", currentContainer.ImageID)
		err = s.dockerClient.RemoveContainerAndImage(ctx, currentContainer)
		if err != nil {
			return fmt.Errorf("error removing image: %w", err)
		}
	} else {
		slog.Info("No container found for image, assuming a fresh deployment...", "image_name", input.ImageName)
	}

	// Pull the new image
	slog.Info("Pulling new image", "image_name", input.ImageName, "tag", input.Tag)
	err = s.dockerClient.PullImageWithAuth(ctx, input.ImageName, input.Tag)
	if err != nil {
		return fmt.Errorf("error pulling image: %w", err)
	}

	// Start new container
	conf := ServiceConfig(input)
	opts := docker.RunImageOptions{
		ContainerName:  _serviceName,
		Config:         conf,
		HostConfig:     nil,
		NetworkingConf: defaultNetworkingConfig,
	}
	newContainerID, err := s.dockerClient.RunImageWithOpts(
		ctx,
		opts,
	)
	if err != nil {
		return fmt.Errorf("error running image %s: %w", input.ImageName, err)
	}
	slog.Info("Started new container", "container_id", newContainerID)

	return nil
}

func (s *SimpleDeployment) runTraefik(ctx context.Context, input *models.DeployInput) (string, error) {
	// TODO: refactor the handling of running images and rerunning them
	// Check if Traefik is already running by image name
	currentTraefik, err := s.dockerClient.GetContainerByImageName(ctx, "traefik")
	if err != nil {
		return "", fmt.Errorf("error getting Traefik container: %w", err)
	}

	if currentTraefik != nil {
		// Extract current version from image tag
		currentVersion := strings.Split(currentTraefik.Image, ":")[1]
		requestedVersion := strings.Split(*input.TraefikImage, ":")[1]

		if currentVersion != requestedVersion {
			slog.Info("Stopping old Traefik version", "current_version", currentVersion, "new_version", requestedVersion)
			err = s.dockerClient.StopContainer(ctx, currentTraefik.ID)
			if err != nil {
				return "", fmt.Errorf("error stopping old Traefik container: %w", err)
			}
			err = s.dockerClient.RemoveContainerAndImage(ctx, currentTraefik)
			if err != nil {
				return "", fmt.Errorf("error removing old Traefik container and image: %w", err)
			}
		} else {
			slog.Info("Traefik already running at requested version", "version", currentVersion)
			return currentTraefik.ID, nil
		}
	}

	if input.TraefikImage == nil || input.TraefikDomainName == nil || input.TraefikACMEEmail == nil {
		return "", errors.New("missing required Traefik configuration")
	}

	// Pull the Traefik image
	imageAndTag := strings.Split(*input.TraefikImage, ":")
	if len(imageAndTag) != 2 { //nolint:mnd // we just expected an image:tag
		return "", fmt.Errorf("invalid Traefik image: %s", *input.TraefikImage)
	}

	err = s.dockerClient.PullImageNoAuth(ctx, imageAndTag[0], imageAndTag[1])
	if err != nil {
		return "", fmt.Errorf("error pulling Traefik image: %w", err)
	}

	traefikConfig := traefik.Config{
		NetworkName:  _defaultNetworkName,
		DomainName:   *input.TraefikDomainName,
		TraefikImage: *input.TraefikImage,
		ACMEEmail:    *input.TraefikACMEEmail,
	}
	dockerConfig := traefik.GetDockerConfig(traefikConfig)

	opts := docker.RunImageOptions{
		ContainerName:  _traefikName,
		Config:         dockerConfig.ContainerConfig,
		HostConfig:     dockerConfig.HostConfig,
		NetworkingConf: defaultNetworkingConfig,
	}
	containerID, err := s.dockerClient.RunImageWithOpts(
		ctx,
		opts,
	)
	if err != nil {
		return "", fmt.Errorf("error running Traefik container: %w", err)
	}
	return containerID, nil
}

func waitForTraefik(ctx context.Context, dockerClient docker.Client, traefikContainerID string) error {
	slog.Info("waiting for Traefik to start", "container_id", traefikContainerID)
	// TODO: configure timeouts
	// TODO: try using the /ping endpoint instead
	// Maximum number of retries
	maxRetries := 30 // 30 seconds timeout

	lastState := ""
	for i := 0; i < maxRetries; i++ {
		container, err := dockerClient.GetContainerByID(ctx, traefikContainerID)
		slog.Info("waiting for Traefik to start", "container", container, "lastState", lastState)
		if err != nil {
			return fmt.Errorf("error inspecting Traefik container: %w", err)
		}
		if container != nil {
			if container.State == "running" {
				slog.Info("Traefik is running", "container_id", traefikContainerID)
				return nil
			}
			lastState = container.State
		}

		// Wait for 1 second before next check
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(time.Second):
			continue
		}
	}

	return fmt.Errorf("timeout waiting for Traefik to start, last state: %s", lastState)
}

func ServiceConfig(input *models.DeployInput) *container.Config {
	newImage := fmt.Sprintf("%s:%s", input.ImageName, input.Tag)

	labels := map[string]string{
		"traefik.enable":                              "true",
		"traefik.http.routers.bondi.rule":             fmt.Sprintf("Host(`%[1]s`) || Host(`www.%[1]s`)", *input.TraefikDomainName),
		"traefik.http.routers.bondi.entrypoints":      "websecure",
		"traefik.http.routers.bondi.tls":              "true",
		"traefik.http.routers.bondi.tls.certresolver": "bondi_resolver",
	}

	env := []string{}
	for k, v := range input.EnvVars {
		env = append(env, fmt.Sprintf("%s=%s", k, v))
	}

	// Create exposed ports configuration
	exposedPorts := nat.PortSet{
		nat.Port(fmt.Sprintf("%d/tcp", input.Port)): {},
	}

	conf := container.Config{
		Image:        newImage,
		Env:          env,
		Labels:       labels,
		ExposedPorts: exposedPorts,
	}

	return &conf
}

File: lib/server/strategy/simple.ml
open Ppx_yojson_conv_lib.Yojson_conv
open Json_helpers

let ( let* ) = Result.bind
let default_network_name = "bondi-network"
let service_name = "bondi-workload"
let traefik_name = "bondi-traefik"

(* ------------------------------------------------------------------------- *)
(* Types                                                                     *)
(* ------------------------------------------------------------------------- *)

type cron_job = {
  name : string;
  image : string;
  schedule : string;
  env_vars : string_map option;
  registry_user : string option; [@default None]
  registry_pass : string option; [@default None]
}
[@@deriving yojson]

type deploy_input = {
  image : string; (* Full image string including tag *)
  port : int;
  registry_user : string option;
  registry_pass : string option;
  env_vars : string_map option;
  traefik_domain_name : string option;
  traefik_image : string option;
  traefik_acme_email : string option;
  force_traefik_redeploy : bool option;
  cron_jobs : cron_job list option; [@default None]
}
[@@deriving yojson]

type deploy_context = {
  current_traefik : Docker.Client.container option;
  current_workload : Docker.Client.container option;
}

type action =
  | CreateNetwork of { network_name : string }
  | EnsureTraefik of {
      image : string;
      domain_name : string;
      acme_email : string;
      traefik_config : Docker.Traefik.docker_config;
    }
  | StopAndRemoveContainer of Docker.Client.container
  | PullImage of { image : string; tag : string; with_auth : bool }
  | RunWorkload of {
      container_name : string;
      config : Docker.Client.container_config;
      networking_conf : Docker.Client.networking_config;
    }

(* ------------------------------------------------------------------------- *)
(* Pure helpers                                                              *)
(* ------------------------------------------------------------------------- *)

let default_networking_config : Docker.Client.networking_config =
  let endpoint : Docker.Client.endpoint_config =
    { aliases = None; ipv4_address = None }
  in
  { endpoints_config = Some [ (default_network_name, endpoint) ] }

let env_vars_to_list env_vars =
  List.map (fun (key, value) -> key ^ "=" ^ value) env_vars

let parse_image_and_tag image =
  match String.split_on_char ':' image with
  | [ name; tag ] -> Ok (name, tag)
  | [ name ] -> Ok (name, "")
  | _ -> Error ("invalid image format: " ^ image)

let service_config (input : deploy_input) :
    (Docker.Client.container_config, string) result =
  match input.traefik_domain_name with
  | None -> Error "missing traefik_domain_name for service labels"
  | Some domain_name ->
      let new_image =
        match parse_image_and_tag input.image with
        | Ok (name, tag) -> if tag = "" then name ^ ":latest" else input.image
        | Error _ -> input.image
      in
      let labels : string_map =
        [
          ("traefik.enable", "true");
          ( "traefik.http.routers.bondi.rule",
            Printf.sprintf "Host(`%s`) || Host(`www.%s`)" domain_name
              domain_name );
          ("traefik.http.routers.bondi.entrypoints", "websecure");
          ("traefik.http.routers.bondi.tls", "true");
          ("traefik.http.routers.bondi.tls.certresolver", "bondi_resolver");
        ]
      in
      let env = Option.map env_vars_to_list input.env_vars in
      let config : Docker.Client.container_config =
        {
          image = Some new_image;
          env;
          cmd = None;
          entrypoint = None;
          hostname = None;
          working_dir = None;
          labels = Some labels;
          exposed_ports = None;
        }
      in
      Ok config

(* ------------------------------------------------------------------------- *)
(* Phase 1: Gather context (read-only)                                       *)
(* ------------------------------------------------------------------------- *)

let image_name_for_lookup image =
  match parse_image_and_tag image with
  | Ok (name, _) -> name
  | Error _ -> image

let gather_context ~client ~net (input : deploy_input) :
    (deploy_context, string) result =
  try
    let current_traefik =
      Docker.Client.get_container_by_image_name client ~net
        ~image_name:"traefik"
    in
    let current_workload =
      Docker.Client.get_container_by_image_name client ~net
        ~image_name:(image_name_for_lookup input.image)
    in
    Ok { current_traefik; current_workload }
  with
  | exn -> Error (Printexc.to_string exn)

(* ------------------------------------------------------------------------- *)
(* Phase 2: Plan (pure)                                                     *)
(* ------------------------------------------------------------------------- *)

let should_run_traefik (input : deploy_input) =
  Option.is_some input.traefik_domain_name
  && Option.is_some input.traefik_image
  && Option.is_some input.traefik_acme_email

let should_redeploy_traefik (input : deploy_input)
    (current_traefik : Docker.Client.container) : bool =
  let force_redeploy =
    Option.value ~default:false input.force_traefik_redeploy
  in
  if force_redeploy then true
  else
    match (input.traefik_image, parse_image_and_tag current_traefik.image) with
    | None, _ -> false
    | Some requested_image, Ok (_name, current_tag) -> (
        match parse_image_and_tag requested_image with
        | Error _ -> false
        | Ok (_requested_name, requested_tag) -> current_tag <> requested_tag)
    | _ -> false

let plan (input : deploy_input) (context : deploy_context) :
    (action list, string) result =
  let actions = ref [] in
  (* Traefik path *)
  let* () =
    if should_run_traefik input then (
      actions :=
        CreateNetwork { network_name = default_network_name } :: !actions;
      let need_traefik_deploy =
        match context.current_traefik with
        | None -> true
        | Some container ->
            if should_redeploy_traefik input container then (
              actions := StopAndRemoveContainer container :: !actions;
              true)
            else false
      in
      if need_traefik_deploy then
        match
          ( input.traefik_image,
            input.traefik_domain_name,
            input.traefik_acme_email )
        with
        | Some image, Some domain_name, Some acme_email -> (
            match parse_image_and_tag image with
            | Error msg -> Error msg
            | Ok (image_name, _image_tag) ->
                let traefik_config : Docker.Traefik.config =
                  {
                    network_name = default_network_name;
                    domain_name;
                    traefik_image = Some image;
                    acme_email;
                  }
                in
                let docker_config =
                  Docker.Traefik.get_docker_config traefik_config
                in
                actions :=
                  EnsureTraefik
                    {
                      image = image_name;
                      domain_name;
                      acme_email;
                      traefik_config = docker_config;
                    }
                  :: !actions;
                Ok ())
        | _ -> Error "missing required traefik configuration"
      else Ok ())
    else Ok ()
  in
  (* Workload path - only when we have a service (traefik_domain_name) *)
  (match input.traefik_domain_name with
  | None -> ()
  | Some _ -> (
      match service_config input with
      | Error _ -> ()
      | Ok service_cfg ->
          (match context.current_workload with
          | None -> ()
          | Some container ->
              actions := StopAndRemoveContainer container :: !actions);
          (match parse_image_and_tag input.image with
          | Ok (image_name, image_tag) ->
              actions :=
                PullImage
                  {
                    image = image_name;
                    tag = (if image_tag = "" then "latest" else image_tag);
                    with_auth = Option.is_some input.registry_user;
                  }
                :: !actions
          | Error _ -> ());
          actions :=
            RunWorkload
              {
                container_name = service_name;
                config = service_cfg;
                networking_conf = default_networking_config;
              }
            :: !actions));
  Ok (List.rev !actions)

(* ------------------------------------------------------------------------- *)
(* Phase 3: Interpreter                                                     *)
(* ------------------------------------------------------------------------- *)

let wait_for_traefik ~clock ~client ~net ~container_id : (unit, string) result =
  let max_retries = 30 in
  let rec loop attempt last_state =
    if attempt >= max_retries then
      Error
        (Printf.sprintf "timeout waiting for Traefik to start, last state: %s"
           last_state)
    else
      let state =
        try
          let inspect =
            Docker.Client.inspect_container client ~net ~container_id
          in
          Ok inspect.state.status
        with
        | exn -> Error (Printexc.to_string exn)
      in
      match state with
      | Ok "running" -> Ok ()
      | Ok "created" ->
          Docker.Client.start_container client ~net ~container_id;
          Eio.Time.sleep clock 1.0;
          loop (attempt + 1) "created"
      | Ok "exited" ->
          Docker.Client.start_container client ~net ~container_id;
          Eio.Time.sleep clock 1.0;
          loop (attempt + 1) "exited"
      | Ok status ->
          Eio.Time.sleep clock 1.0;
          loop (attempt + 1) status
      | Error msg -> Error msg
  in
  loop 0 ""

let interpret ~clock ~client ~net (actions : action list) :
    (unit, string) result =
  let rec run = function
    | [] -> Ok ()
    | CreateNetwork { network_name } :: rest ->
        Docker.Client.create_network_if_not_exists client ~net ~network_name;
        run rest
    | EnsureTraefik { image; traefik_config; domain_name = _; acme_email = _ }
      :: rest ->
        let full_image =
          match traefik_config.container_config.image with
          | Some img -> img
          | None -> image
        in
        let image_name, image_tag =
          match parse_image_and_tag full_image with
          | Ok (n, t) -> (n, t)
          | Error _ -> (image, "")
        in
        Docker.Client.pull_image_no_auth client ~net ~image:image_name
          ~tag:image_tag;
        let opts : Docker.Client.run_image_options =
          {
            container_name = traefik_name;
            config = traefik_config.container_config;
            host_config = Some traefik_config.host_config;
            networking_conf = Some default_networking_config;
          }
        in
        let container_id = Docker.Client.run_image_with_opts client ~net opts in
        let* () = wait_for_traefik ~clock ~client ~net ~container_id in
        run rest
    | StopAndRemoveContainer container :: rest ->
        Docker.Client.stop_container client ~net ~container_id:container.id;
        Docker.Client.remove_container_and_image client ~net ~container;
        run rest
    | PullImage { image; tag; with_auth } :: rest ->
        if with_auth then
          Docker.Client.pull_image_with_auth client ~net ~image ~tag
        else Docker.Client.pull_image_no_auth client ~net ~image ~tag;
        run rest
    | RunWorkload { container_name; config; networking_conf } :: rest ->
        let opts : Docker.Client.run_image_options =
          {
            container_name;
            config;
            host_config = None;
            networking_conf = Some networking_conf;
          }
        in
        let _ = Docker.Client.run_image_with_opts client ~net opts in
        run rest
  in
  run actions

(* ------------------------------------------------------------------------- *)
(* Entry point                                                               *)
(* ------------------------------------------------------------------------- *)

let deploy ~clock ~client ~net (input : deploy_input) : (unit, string) result =
  let* context = gather_context ~client ~net input in
  let* actions = plan input context in
  interpret ~clock ~client ~net actions
